Fix Blinking Sensor problem

petecressman opened the pull request:
As long as the animated icon is not rotated, the animation will function.  Reset the rotation angle to 0 degrees to restore a broken blinking icon.  As far as I can tell, to make animated icons work after rotation, the gif file would need to be disassembled, the individual frame rotated and then reassemble the file.  Too much work for the time I have available.   All the previous decoration features for Sensors, Memories and Labels still work.  I hope this is satisfactory.

bobjacobsen commented:
Am looking at this now.

bobjacobsen commented:
There's a conflict between this patch and the current release-4.2 branch in java/src/jmri/jmrit/display/configurexml/PositionableLabelXml.java near line 394:

````
<<<<<<< HEAD
        if (deg == 0 && util.hasBackground()) {
            l.setOpaque(true);
=======
        l.rotate(deg);
        if (deg == 0 && util.hasBackground()) {
            l.setOpaque(true);                    
>>>>>>> 25d19a6c120074f1cb892890b1f6f7bd9ea0c71a
````

the l.rotate(deg); is a deliberate addition in the patch, but I don't (yet) see what it does. Looking...

rhwood commented on java/test/jmri/jmrit/display/MemoryIconTest.java:
```diff
@@ -91,6 +91,7 @@ protected void init() {
         Assert.assertNotNull("Instance exists", i);
         jmri.InstanceManager.memoryManagerInstance().provideMemory("IM2").setValue("");
         to.setMemory("IM2");
+        to.setMemory("IM2");
```
This test does not appear to be correct.

bobjacobsen replied:
This is reverting something that was fixed on master a week or so ago.  This makes me suspect what's been included in these two commits...

bobjacobsen replied:
More specifically, Pete added the duplicate setMemory in 1b731bbc481 14 days ago, and I fixed it in 5e949927 and bb529913ea1509.

bobjacobsen replied:
Note that the actual 25d19a6 commit isn't bringing in this change; it's coming in through the previous 7fd2d48 commit which is commented as 
```
Merge branch 'master' of https://github.com/JMRI/JMRI
```
but seems to be carrying non-master changes from its two parents.

rhwood commented:
Rotating an animated gif appears to be fairly simple. http://stackoverflow.com/questions/26005726/rotate-an-animated-gif-imageicon-using-affinetransform

rhwood commented:
@petecressman Please update the master branch in your fork with the current JMRI master.

petecressman commented:
l.rotate(deg); needs to be called to set the rotation of the icon.  This is for singlet icons, not Sensor icon (they override and handle it differently).   Perhaps the line could be executed at 389 only, but passing through rotate(0) sets some flags.

The conditional to setOpaque(true); is essential.

This code is so fragile, the cases that need visual inspection are numerous.   The monkey wrench in the soup is Sensor icons that can be both icons and text ("Text Overlaid").  My opinion is this is a marginal feature at best and can be simulated with a Label item superimposed on a sensor icon.  Nevertheless

I added the feature rotation to an arbitrary angle and felt obligated to insure that all the previous decoration stuff rotated as well.   I feel the rotation of JLabel classes is a valuable feature, less so for the decorations, but they are nice to have. But sensors having three modes of display is too much.

 

Pete C.

petecressman commented:
l.rotate(deg); needs to be called to set the rotation of the icon.  This is for singlet icons, nor Sensor icon (they override and handle it differently).   Perhaps the line could be executed at 389 only, but passing through rotate(0) sets some flags.
The conditional to setOpaque(true); is essential.
This code is so fragile, the cases that need visual inspection are numerous.   The monkey wrench in the soup is Sensor icons that can be both icons and text ("Text Overlaid").  My opinion is this is a marginal feature at best and can be simulated with a Label item superimposed on a sensor icon.  Nevertheless

I added the feature rotation to an arbitrary angle and felt obligated to insure that all the previous decoration stuff rotated as well.   I feel the rotation of JLabel classes is a valuable feature, less so for the decorations, but they are nice to have. But sensors having three modes of display is too much.
 Pete C. 

    On Wednesday, December 23, 2015 7:19 AM, Bob Jacobsen <notifications@github.com> wrote:
 

 Am looking at this now.—
Reply to this email directly or view it on GitHub.

bobjacobsen commented:
Some of the content in this PR was included in JMRI 4.2 see discussion in #316

DanielBoudreau commented:
When are we going to pull this into head?  I understand that there are conflicts, is Pete going to address them?

petecressman commented:
I'm having lots of problems pulling from the head using Eclipse to see and resolve the conflicts.  I upgraded from windows 7 to 10 a few weeks ago.  And now....
1. Pull from eclipse says there is a conflict in the tree to file but nothing else is is shown. 
2. Opened the Git Shell and the message was to update Git.3. Did the update - message was it will take time. When i returned to the computer command prompt was there with message to set environment path variable.4. Did that.  Now in Git Shell, git commands don't find a repository.  e.g.
C:\Users\SVD\Documents\GitHub> git remote -v
fatal: Not a git repository (or any of the parent directories): .git
Also now from Eclipse, Pull has message 
Cannot pull into a repository with state MERGINGWrongRepositoryStateException
5.  Deleted file said to be in conflict and replaced it.  Now Eclipse has a red dot on the file. (clean build still works)
Everything i have done seems to have made things worse.Don't pull whatever is in my fork.
What would be the best way to restart?
Pete
 

    On Sunday, December 27, 2015 10:53 AM, Daniel Boudreau <notifications@github.com> wrote:
 

 When are we going to pull this into head? I understand that there are conflicts, is Pete going to address them?—
Reply to this email directly or view it on GitHub.

DanielBoudreau commented:
Hi Pete,

 

Do you have anything on your Eclipse machine that you need to commit?

 

I’m thinking that if you might need to do a complete replace.  You can create a new workspace and do a clean pull. Then from your other workspace grab any files that you need to commit if you have any.

 

What I’ve found is that the only way to get rid of conflicts in Eclipse is to replace the file from head which then removes all changes that you’ve done.

 

I never use the git shell, only Eclipse commands.

 

Dan

 

From: petecressman [mailto:notifications@github.com] 
Sent: Sunday, December 27, 2015 3:18 PM
To: JMRI/JMRI <JMRI@noreply.github.com>
Cc: Daniel Boudreau <daboudreau@hotmail.com>
Subject: Re: [JMRI] Fix Blinking Sensor problem (#394)

 

I'm having lots of problems pulling from the head using Eclipse to see and resolve the conflicts.  I upgraded from windows 7 to 10 a few weeks ago.  And now....
1. Pull from eclipse says there is a conflict in the tree to file but nothing else is is shown. 
2. Opened the Git Shell and the message was to update Git.3. Did the update - message was it will take time. When i returned to the computer command prompt was there with message to set environment path variable.4. Did that.  Now in Git Shell, git commands don't find a repository.  e.g.
C:\Users\SVD\Documents\GitHub> git remote -v
fatal: Not a git repository (or any of the parent directories): .git
Also now from Eclipse, Pull has message 
Cannot pull into a repository with state MERGINGWrongRepositoryStateException
5.  Deleted file said to be in conflict and replaced it.  Now Eclipse has a red dot on the file. (clean build still works)
Everything i have done seems to have made things worse.Don't pull whatever is in my fork.
What would be the best way to restart?
Pete


On Sunday, December 27, 2015 10:53 AM, Daniel Boudreau <notifications@github.com <mailto:notifications@github.com> > wrote:


When are we going to pull this into head? I understand that there are conflicts, is Pete going to address them?—
Reply to this email directly or view it on GitHub. 




—
Reply to this email directly or view it on GitHub <https://github.com/JMRI/JMRI/pull/394#issuecomment-167439341> .  <https://github.com/notifications/beacon/ANw2p01-jqUPir2Fi6uCDqkIATmRmjLpks5pUD7jgaJpZM4G6Uy1.gif>

petecressman commented:
Hi Dan,
I'm downloading a new work space now.  Do you think I can just copy in and replace the files I need from the broken workspace - Or is line by line editing necessary?
Thanks,Pete
 

    On Sunday, December 27, 2015 12:31 PM, Daniel Boudreau <notifications@github.com> wrote:
 

 Hi Pete,

 

Do you have anything on your Eclipse machine that you need to commit?

 

I’m thinking that if you might need to do a complete replace. You can create a new workspace and do a clean pull. Then from your other workspace grab any files that you need to commit if you have any.

 

What I’ve found is that the only way to get rid of conflicts in Eclipse is to replace the file from head which then removes all changes that you’ve done.

 

I never use the git shell, only Eclipse commands.

 

Dan

 

From: petecressman [mailto:notifications@github.com] 
Sent: Sunday, December 27, 2015 3:18 PM
To: JMRI/JMRI <JMRI@noreply.github.com>
Cc: Daniel Boudreau <daboudreau@hotmail.com>
Subject: Re: [JMRI] Fix Blinking Sensor problem (#394)

 

I'm having lots of problems pulling from the head using Eclipse to see and resolve the conflicts. I upgraded from windows 7 to 10 a few weeks ago. And now....
1. Pull from eclipse says there is a conflict in the tree to file but nothing else is is shown. 
2. Opened the Git Shell and the message was to update Git.3. Did the update - message was it will take time. When i returned to the computer command prompt was there with message to set environment path variable.4. Did that. Now in Git Shell, git commands don't find a repository. e.g.
C:\Users\SVD\Documents\GitHub> git remote -v
fatal: Not a git repository (or any of the parent directories): .git
Also now from Eclipse, Pull has message 
Cannot pull into a repository with state MERGINGWrongRepositoryStateException
5. Deleted file said to be in conflict and replaced it. Now Eclipse has a red dot on the file. (clean build still works)
Everything i have done seems to have made things worse.Don't pull whatever is in my fork.
What would be the best way to restart?
Pete


On Sunday, December 27, 2015 10:53 AM, Daniel Boudreau <notifications@github.com <mailto:notifications@github.com> > wrote:


When are we going to pull this into head? I understand that there are conflicts, is Pete going to address them?—
Reply to this email directly or view it on GitHub. 




—
Reply to this email directly or view it on GitHub <https://github.com/JMRI/JMRI/pull/394#issuecomment-167439341> . <https://github.com/notifications/beacon/ANw2p01-jqUPir2Fi6uCDqkIATmRmjLpks5pUD7jgaJpZM4G6Uy1.gif> 

—
Reply to this email directly or view it on GitHub.

DanielBoudreau commented:
I’ve been able to paste in files.  I would also replace your fork on github with a new one.  This way you’re starting out clean.

 

Dan

 

From: petecressman [mailto:notifications@github.com] 
Sent: Sunday, December 27, 2015 3:37 PM
To: JMRI/JMRI <JMRI@noreply.github.com>
Cc: Daniel Boudreau <daboudreau@hotmail.com>
Subject: Re: [JMRI] Fix Blinking Sensor problem (#394)

 

Hi Dan,
I'm downloading a new work space now.  Do you think I can just copy in and replace the files I need from the broken workspace - Or is line by line editing necessary?
Thanks,Pete


On Sunday, December 27, 2015 12:31 PM, Daniel Boudreau <notifications@github.com <mailto:notifications@github.com> > wrote:


Hi Pete,



Do you have anything on your Eclipse machine that you need to commit?



I’m thinking that if you might need to do a complete replace. You can create a new workspace and do a clean pull. Then from your other workspace grab any files that you need to commit if you have any.



What I’ve found is that the only way to get rid of conflicts in Eclipse is to replace the file from head which then removes all changes that you’ve done.



I never use the git shell, only Eclipse commands.



Dan



From: petecressman [mailto:notifications@github.com] 
Sent: Sunday, December 27, 2015 3:18 PM
To: JMRI/JMRI <JMRI@noreply.github.com <mailto:JMRI@noreply.github.com> >
Cc: Daniel Boudreau <daboudreau@hotmail.com <mailto:daboudreau@hotmail.com> >
Subject: Re: [JMRI] Fix Blinking Sensor problem (#394)



I'm having lots of problems pulling from the head using Eclipse to see and resolve the conflicts. I upgraded from windows 7 to 10 a few weeks ago. And now....
1. Pull from eclipse says there is a conflict in the tree to file but nothing else is is shown. 
2. Opened the Git Shell and the message was to update Git.3. Did the update - message was it will take time. When i returned to the computer command prompt was there with message to set environment path variable.4. Did that. Now in Git Shell, git commands don't find a repository. e.g.
C:\Users\SVD\Documents\GitHub> git remote -v
fatal: Not a git repository (or any of the parent directories): .git
Also now from Eclipse, Pull has message 
Cannot pull into a repository with state MERGINGWrongRepositoryStateException
5. Deleted file said to be in conflict and replaced it. Now Eclipse has a red dot on the file. (clean build still works)
Everything i have done seems to have made things worse.Don't pull whatever is in my fork.
What would be the best way to restart?
Pete


On Sunday, December 27, 2015 10:53 AM, Daniel Boudreau <notifications@github.com  <mailto:notifications@github.com%20%3cmailto:notifications@github.com> <mailto:notifications@github.com> > wrote:


When are we going to pull this into head? I understand that there are conflicts, is Pete going to address them?—
Reply to this email directly or view it on GitHub. 




—
Reply to this email directly or view it on GitHub <https://github.com/JMRI/JMRI/pull/394#issuecomment-167439341> . <https://github.com/notifications/beacon/ANw2p01-jqUPir2Fi6uCDqkIATmRmjLpks5pUD7jgaJpZM4G6Uy1.gif> 

—
Reply to this email directly or view it on GitHub. 




—
Reply to this email directly or view it on GitHub <https://github.com/JMRI/JMRI/pull/394#issuecomment-167440086> .  <https://github.com/notifications/beacon/ANw2p3Ehi8CjQe1mnR7uYGRKYUhyJcNrks5pUEN-gaJpZM4G6Uy1.gif>

rhwood commented:
I would do a line-by-line replacement *after* syncing from both your fork and JMRI/master.

I found that all IDEs are somewhat weak with Forks and Git. I would recommend using GitHub Desktop (easier) or SourceTree (seriously powerful and complex) to manage git syncing and merging, and your IDE for coding only.

Randall Wood
Alexandria Software
randall.h.wood@alexandriasoftware.com
http://alexandriasoftware.com

> On Dec 27, 2015, at 15:37, petecressman <notifications@github.com> wrote:
> 
> Hi Dan,
> I'm downloading a new work space now.  Do you think I can just copy in and replace the files I need from the broken workspace - Or is line by line editing necessary?
> Thanks,Pete
> 
> 
> On Sunday, December 27, 2015 12:31 PM, Daniel Boudreau <notifications@github.com> wrote:
> 
> 
> Hi Pete,
> 
> 
> 
> Do you have anything on your Eclipse machine that you need to commit?
> 
> 
> 
> I’m thinking that if you might need to do a complete replace. You can create a new workspace and do a clean pull. Then from your other workspace grab any files that you need to commit if you have any.
> 
> 
> 
> What I’ve found is that the only way to get rid of conflicts in Eclipse is to replace the file from head which then removes all changes that you’ve done.
> 
> 
> 
> I never use the git shell, only Eclipse commands.
> 
> 
> 
> Dan
> 
> 
> 
> From: petecressman [mailto:notifications@github.com] 
> Sent: Sunday, December 27, 2015 3:18 PM
> To: JMRI/JMRI <JMRI@noreply.github.com>
> Cc: Daniel Boudreau <daboudreau@hotmail.com>
> Subject: Re: [JMRI] Fix Blinking Sensor problem (#394)
> 
> 
> 
> I'm having lots of problems pulling from the head using Eclipse to see and resolve the conflicts. I upgraded from windows 7 to 10 a few weeks ago. And now....
> 1. Pull from eclipse says there is a conflict in the tree to file but nothing else is is shown. 
> 2. Opened the Git Shell and the message was to update Git.3. Did the update - message was it will take time. When i returned to the computer command prompt was there with message to set environment path variable.4. Did that. Now in Git Shell, git commands don't find a repository. e.g.
> C:\Users\SVD\Documents\GitHub> git remote -v
> fatal: Not a git repository (or any of the parent directories): .git
> Also now from Eclipse, Pull has message 
> Cannot pull into a repository with state MERGINGWrongRepositoryStateException
> 5. Deleted file said to be in conflict and replaced it. Now Eclipse has a red dot on the file. (clean build still works)
> Everything i have done seems to have made things worse.Don't pull whatever is in my fork.
> What would be the best way to restart?
> Pete
> 
> 
> On Sunday, December 27, 2015 10:53 AM, Daniel Boudreau <notifications@github.com <mailto:notifications@github.com> > wrote:
> 
> 
> When are we going to pull this into head? I understand that there are conflicts, is Pete going to address them?—
> Reply to this email directly or view it on GitHub. 
> 
> 
> 
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/JMRI/JMRI/pull/394#issuecomment-167439341> . <https://github.com/notifications/beacon/ANw2p01-jqUPir2Fi6uCDqkIATmRmjLpks5pUD7jgaJpZM4G6Uy1.gif> 
> 
> —
> Reply to this email directly or view it on GitHub. 
> 
> —
> Reply to this email directly or view it on GitHub.
>

rhwood commented:
Don't replace your fork in GitHub, it's still good and clean. Check out (or clone) your fork and then sync in JMRI/Master (GitHub Desktop is really good at handling this).

Randall Wood
Alexandria Software
randall.h.wood@alexandriasoftware.com
http://alexandriasoftware.com

> On Dec 27, 2015, at 15:42, Daniel Boudreau <notifications@github.com> wrote:
> 
> I’ve been able to paste in files. I would also replace your fork on github with a new one. This way you’re starting out clean.
> 
> 
> 
> Dan
> 
> 
> 
> From: petecressman [mailto:notifications@github.com] 
> Sent: Sunday, December 27, 2015 3:37 PM
> To: JMRI/JMRI <JMRI@noreply.github.com>
> Cc: Daniel Boudreau <daboudreau@hotmail.com>
> Subject: Re: [JMRI] Fix Blinking Sensor problem (#394)
> 
> 
> 
> Hi Dan,
> I'm downloading a new work space now. Do you think I can just copy in and replace the files I need from the broken workspace - Or is line by line editing necessary?
> Thanks,Pete
> 
> 
> On Sunday, December 27, 2015 12:31 PM, Daniel Boudreau <notifications@github.com <mailto:notifications@github.com> > wrote:
> 
> 
> Hi Pete,
> 
> 
> 
> Do you have anything on your Eclipse machine that you need to commit?
> 
> 
> 
> I’m thinking that if you might need to do a complete replace. You can create a new workspace and do a clean pull. Then from your other workspace grab any files that you need to commit if you have any.
> 
> 
> 
> What I’ve found is that the only way to get rid of conflicts in Eclipse is to replace the file from head which then removes all changes that you’ve done.
> 
> 
> 
> I never use the git shell, only Eclipse commands.
> 
> 
> 
> Dan
> 
> 
> 
> From: petecressman [mailto:notifications@github.com] 
> Sent: Sunday, December 27, 2015 3:18 PM
> To: JMRI/JMRI <JMRI@noreply.github.com <mailto:JMRI@noreply.github.com> >
> Cc: Daniel Boudreau <daboudreau@hotmail.com <mailto:daboudreau@hotmail.com> >
> Subject: Re: [JMRI] Fix Blinking Sensor problem (#394)
> 
> 
> 
> I'm having lots of problems pulling from the head using Eclipse to see and resolve the conflicts. I upgraded from windows 7 to 10 a few weeks ago. And now....
> 1. Pull from eclipse says there is a conflict in the tree to file but nothing else is is shown. 
> 2. Opened the Git Shell and the message was to update Git.3. Did the update - message was it will take time. When i returned to the computer command prompt was there with message to set environment path variable.4. Did that. Now in Git Shell, git commands don't find a repository. e.g.
> C:\Users\SVD\Documents\GitHub> git remote -v
> fatal: Not a git repository (or any of the parent directories): .git
> Also now from Eclipse, Pull has message 
> Cannot pull into a repository with state MERGINGWrongRepositoryStateException
> 5. Deleted file said to be in conflict and replaced it. Now Eclipse has a red dot on the file. (clean build still works)
> Everything i have done seems to have made things worse.Don't pull whatever is in my fork.
> What would be the best way to restart?
> Pete
> 
> 
> On Sunday, December 27, 2015 10:53 AM, Daniel Boudreau <notifications@github.com <mailto:notifications@github.com%20%3cmailto:notifications@github.com> <mailto:notifications@github.com> > wrote:
> 
> 
> When are we going to pull this into head? I understand that there are conflicts, is Pete going to address them?—
> Reply to this email directly or view it on GitHub. 
> 
> 
> 
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/JMRI/JMRI/pull/394#issuecomment-167439341> . <https://github.com/notifications/beacon/ANw2p01-jqUPir2Fi6uCDqkIATmRmjLpks5pUD7jgaJpZM4G6Uy1.gif> 
> 
> —
> Reply to this email directly or view it on GitHub. 
> 
> 
> 
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/JMRI/JMRI/pull/394#issuecomment-167440086> . <https://github.com/notifications/beacon/ANw2p3Ehi8CjQe1mnR7uYGRKYUhyJcNrks5pUEN-gaJpZM4G6Uy1.gif> 
> 
> —
> Reply to this email directly or view it on GitHub.
>

petecressman commented:
I'm exhausted- I've done 5 downloads today to new workspaces.  All looked OK, but somewhat different from my fork.  I madea few minor changes, nothing substantive.  I have committed then locally, but cannot push to my fork.  Message is
"rejected-non-fast-forward"

Could not find sync in Git Shell.  Did sync from Eclipse and it messed up the entire ws - every file is red marked for error.
With Git Shell unable to get to git running when in a new workspace directory.
Can't do anything with GitHub - just tutorials.

Fork is now 229 commits behind.


sorry, I have no idea what is going on.  If there is a new build someplace I will test it.
